contracts/contracts/strategies/sonic/SonicSwapXAMOStrategy.sol

```diff
@@ -0,0 +1,576 @@
+        uint256 amountOut = IPair(pool).getAmountOut(_amountIn, _tokenIn);
+
+        // Safety check that we are dealing with the correct pool tokens
+        require(
+            (_tokenIn == ws || _tokenIn == os) &&
+                (_tokenOut == ws || _tokenOut == os),
+            "Unsupported swap"
+        );
+
+        // Work out the correct order of the amounts for the pool
+        (uint256 amount0, uint256 amount1) = _tokenIn == ws
+            ? (uint256(0), amountOut)
+            : (amountOut, 0);
+
+        // Perform the swap on the pool
+        IPair(pool).swap(amount0, amount1, address(this), new bytes(0));
+
+        // The slippage protection against the amount out is indirectly done
+        // via the improvePoolBalance and strategyValueChecker modifiers
+    }
+
+    /**
+     * Checks that the protocol is solvent, protecting from a rogue Strategist / Guardian that can
+     * keep rebalancing the pool in both directions making the protocol lose a tiny amount of
+     * funds each time.
+     *
+     * Protocol must be at least SOLVENCY_THRESHOLD (99,8 %) backed in order for the rebalances to
+     * function.
+     */
+    function _solvencyAssert() internal view {
+        uint256 _totalVaultValue = IVault(vaultAddress).totalValue();
+        uint256 _totalOSSupply = IERC20(os).totalSupply();
+
+        if (
+            _totalVaultValue.divPrecisely(_totalOSSupply) < SOLVENCY_THRESHOLD
+        ) {
+            revert("Protocol insolvent");
+        }
+    }
+
+    /***************************************
+                Assets and Rewards
+    ****************************************/
+
+    /**
+     * @notice Collect accumulated SWPx (and other) rewards and send to the Harvester.
+     */
+    function collectRewardTokens()
+        external
+        override
+        onlyHarvester
+        nonReentrant
+    {
+        // Collect SWPx rewards from the gauge
+        IGauge(gauge).getReward();
+
+        _collectRewardTokens();
+    }
+
+    /**
+     * @notice Get the total asset value held in the SwapX pool
+     * @param _asset      Address of the Wrapped S (wS) token
+     * @return balance    Total value of the wS and OS tokens held in the pool
+     */
+    function checkBalance(address _asset)
+        public
+        view
+        override
+        returns (uint256 balance)
+    {
+        require(_asset == ws, "Unsupported asset");
+
+        // wS balance needed here for the balance check that happens from vault during depositing.
+        balance = IERC20(ws).balanceOf(address(this));
+
+        uint256 lpTokens = IGauge(gauge).balanceOf(address(this));
+        if (lpTokens == 0) return balance;
+
+        // Add the strategy’s share of the wS and OS tokens in the SwapX pool.
+        // (pool’s wS reserves + pool’s OS reserves) * strategy’s LP tokens / total supply of pool LP tokens
+        (uint256 wsReserves, uint256 osReserves, ) = IPair(pool).getReserves();
+        balance +=
```

Comment:
🟥 This is easy rekt. See notion.

Reply:
I've reimplemented `checkBalance`. Testing shows the strategy `checkBalance` does not change at all when others swap against the pool.